### PR TITLE
[Typing][C-17] Add type annotations for `python/paddle/distributed/communication/gather.py`

### PR DESCRIPTION
**Pull request by BHmingyang**
### PR Category
User Experience

### PR Types
Improvements

### Description
类型标注
python/paddle/distributed/communication/gather.py

### CODE REVIEW

File: python/paddle/distributed/communication/gather.py [hunk 1]
```diff
@@ -13,11 +13,26 @@
...
+if TYPE_CHECKING:
+    from paddle import Tensor
+    from paddle.base.core import task
+    from paddle.distributed.communication.group import Group
+
 
-def gather(tensor, gather_list=None, dst=0, group=None, sync_op=True):
+def gather(
+    tensor: Tensor,
+    gather_list: list = None,
```
Comment (SigureMo) [inline python/paddle/distributed/communication/gather.py]:
```suggestion
    gather_list: list[Tensor] = None,
```

泛型必须写明参数类型

File: python/paddle/distributed/communication/gather.py [hunk 2]
```diff
@@ -13,11 +13,26 @@
...
+    dst: int = 0,
+    group: Group | None = None,
+    sync_op: bool = True,
+) -> task | None:
```
Comment (SigureMo) [inline python/paddle/distributed/communication/gather.py:35]: 返回值是 Union 的情况优先考虑使用 overload 进行区分，这里可以使用 `sync_op` 来进行区分

Comment (SigureMo) [reply python/paddle/distributed/communication/gather.py:35]: 可参考 `python/paddle/tensor/linalg.py` 的 `lu` 函数

Comment (BHmingyang) [reply python/paddle/distributed/communication/gather.py:35]: @overload
def lu(
    x: Tensor,
    pivot: bool = ...,
    get_infos: Literal[True] = ...,
    name: str | None = ...,
) -> tuple[Tensor, Tensor, Tensor]:
    ...


@overload
def lu(
    x: Tensor, pivot: bool = ..., get_infos: bool = ..., name: str | None = ...
) -> tuple[Tensor, Tensor] | tuple[Tensor, Tensor, Tensor]:
    ...


def lu(
    x, pivot=True, get_infos=False, name=None
) -> tuple[Tensor, Tensor] | tuple[Tensor, Tensor, Tensor]:

是也要像这样写多了def lu()吗？还是把返回值改成tuple就行？

Comment (SigureMo) [reply python/paddle/distributed/communication/gather.py:35]: overload 就是要写多个的

**Review by SigureMo — APPROVED** [review]
LGTMeow

**Pull request merged**